OpenRA.Mods.AS/Traits/Nydus.cs

```diff
@@ -0,0 +1,207 @@
+	class NydusTransportableInfo : ITraitInfo
+	{
+		[VoiceReference]
+		public readonly string Voice = "Action";
+		public readonly string EnterCursor = "enter";
+		public readonly string EnterBlockedCursor = "enter-blocked";
+		public object Create(ActorInitializer init) { return new NydusTransportable(init, this); }
+	}
+
+	class NydusTransportable : IIssueOrder, IResolveOrder, IOrderVoice
+	{
+		readonly NydusTransportableInfo info;
+
+		public NydusTransportable(ActorInitializer init, NydusTransportableInfo info)
+		{
+			this.info = info;
+		}
+
+		public IEnumerable<IOrderTargeter> Orders
+		{
+			get { yield return new NydusTransportOrderTargeter(info); }
+		}
+
+		public Order IssueOrder(Actor self, IOrderTargeter order, Target target, bool queued)
+		{
+			if (order.OrderID != "NydusTransport")
+				return null;
+
+			// You can't enter enemy nydus canal so this will do.
+			if (target.Type == TargetType.FrozenActor)
+				return null;
+
+			return new Order(order.OrderID, self, target, queued) { };
+		}
+
+		// Does the player own enough nydus canals?
+		// For nydus canals to work, the player needs at least 2.
+		// Actually, at least 1 but in that case, the unit goes in and immediately comes out. That's stupid.
+		static bool HasEnoughCanals(Actor self)
+		{
+			var counter = self.Owner.PlayerActor.Trait<NydusCounter>();
+			return counter.Count > 1;
+		}
+
+		static bool IsValidOrder(Actor self, Order order)
+		{
+			// Not targeting a frozen actor
+			if (order.ExtraData == 0 && order.TargetActor == null)
+				return false;
+
+			// The owner must own at least two nydus canals.
+			if (!HasEnoughCanals(self))
+				return false;
+
+			// primary buildings are where entered units exit.
+			// You don't want to put them in the primary building.
+			return !order.TargetActor.IsPrimaryNydusExit();
+		}
+
+		public string VoicePhraseForOrder(Actor self, Order order)
+		{
+			return order.OrderString == "NydusTransport" && IsValidOrder(self, order)
+				? info.Voice : null;
+		}
+
+		public void ResolveOrder(Actor self, Order order)
+		{
+			if (order.OrderString != "NydusTransport" || !IsValidOrder(self, order))
+				return;
+
+			var target = self.ResolveFrozenActorOrder(order, Color.Yellow);
+			if (target.Type != TargetType.Actor)
+				return;
+
+			if (!order.Queued)
+				self.CancelActivity();
+
+			self.SetTargetLine(target, Color.Yellow);
+			self.QueueActivity(new EnterNydus(self, target.Actor, EnterBehaviour.Exit));
+		}
+
+		class NydusTransportOrderTargeter : UnitOrderTargeter
+		{
+			NydusTransportableInfo info;
+
+			public NydusTransportOrderTargeter(NydusTransportableInfo info)
+				: base("NydusTransport", 6, info.EnterCursor, false, true)
+			{
+				this.info = info;
+			}
+
+			public override bool CanTargetActor(Actor self, Actor target, TargetModifiers modifiers, ref string cursor)
+			{
+				if (!target.Info.HasTraitInfo<NydusInfo>())
+					return false;
+
+				// can only enter player owned one. (Not even ally's)
```

Comment:
TBH I'd expose this to a Stance field in the trait.

Yes, that would mean  always allowing allies (for now) but doing so could allow to use the logic for wormholes in NG.